lib/bolt/catalog.rb

```diff
@@ -70,9 +70,38 @@ def compile_catalog(request)
                           bolt_inventory: inv) do
             Puppet.lookup(:pal_current_node).trusted_data = target['trusted']
             pal.with_catalog_compiler do |compiler|
-              # This needs to happen inside the catalog compiler so loaders are initialized for loading
-              vars = Puppet::Pops::Serialization::FromDataConverter.convert(request['plan_vars'])
-              pal.send(:add_variables, compiler.send(:topscope), target['variables'].merge(vars))
+              # Deserializing needs to happen inside the catalog compiler so
+              # loaders are initialized for loading
+              plan_vars = Puppet::Pops::Serialization::FromDataConverter.convert(request['plan_vars'])
+
+              # Facts will be set by the catalog compiler, so we need to ensure
+              # that any plan or target variables with the same name are not
+              # passed into the apply block to avoid a redefinition error.
+              # Filter out plan and target vars separately and raise a Puppet
+              # warning if there are any collisions for either. Puppet warning
+              # is the only way to log a message that will make it back to Bolt
+              # to be printed.
+              pv_collisions, pv_filtered = plan_vars.partition do |k, _|
+                target['facts'].keys.include?(k)
+              end.map(&:to_h)
+              print_pv = pv_collisions.keys.map { |k| "$#{k}" }.join(', ')
+              unless print_pv.empty?
+                plural = pv_collisions.keys.length == 1 ? '' : 's'
+                Puppet.warning("Plan variable#{plural} #{print_pv} will be overridden by fact#{plural} " \
+                               "of the same name in the apply block")
+              end
+
+              tv_collisions, tv_filtered = target['variables'].partition do |k, _|
+                target['facts'].keys.include?(k)
+              end.map(&:to_h)
+              print_tv = tv_collisions.keys.map { |k| "$#{k}" }.join(', ')
+              unless print_tv.empty?
+                plural = tv_collisions.keys.length == 1 ? '' : 's'
+                Puppet.warning("Target variable#{plural} #{print_tv} for target #{target['name']} " \
```

Comment:
Bolt automatically prepends these log messages with the target name, so we don't need that part here.